### ISSUE TITLE
[init-declarations]: new options ignorePattern 

### ISSUE THREAD

**Issue by anikethsaha** [eslint/eslint#13107]
**What rule do you want to change?**
`init-declarations`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
As a new options, namely `ignorePattern` or `allow` which accepts an array of regex.
It will simply ignore the literals whose name matches the pattern

**Please provide some example code that this change will affect:**

```js
/*eslint init-declarations: ["error", "always",{ ignorePattern : ['bar'] }]*/

function foo() {
  var bar; // lint free
  let baz; // error
}
```

```js
/*eslint init-declarations: ["error", "never", { ignorePattern : ['bar'] }]*/

function foo() {
  var bar = "string"; // lint free
  let baz = "another-string"; // error
}
```

```js
```

**What does the rule currently do for this code?**
It reports them

**What will the rule do after it's changed?**
It will simple ignore them

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by kaicataldo** [comment]
Can you explain what problem you are trying to solve with this proposal? Since the rule is intended to make the initialization of variables consistent, having exceptions seems to go against the purpose of the rule. I'm also not able to come up with any cases where an exception would be useful, but would love to learn about cases if they exist.

**Comment by anikethsaha** [comment]
One issue I faced is as for `never` options this rule disallows initialization for `var`. 
I have worked in a codebase that uses es5 completely and doesn't count for `const` and `let`.
And there we represent `const` as either in CAPS or something like this `_const_<variable>` name. 
So, I thought it might be good to have a ignore list for this as const variables are pretty much static in it and we can pick them for a single file. Might be helpful for future similar cases

**Comment by kaicataldo** [comment]
This sounds like a case where it would be best to just turn off the rule, since I can't think of any reasons why one would _need_ an escape hatch. Please let me know if I'm missing something, though!

**Comment by anikethsaha** [comment]
turning it off won't check for other variables. 
We had a pattern for constants (UPPERCASE) which I wanted to initialized but not the other ones.

**Comment by kaicataldo** [comment]
To explain my thought process a bit more, this rule is currently a fairly straightforward stylistic rule that’s stated intention is “to bring consistency to variable initializations and declarations.” While adding extra configuration increases flexibility, it also makes this simple rule significantly more complex. This specific proposal also seems to go against the intention of the rule, which I’m generally opposed to.

Do you mind explaining what problem you’re trying to solve with this proposal? Apologies if I’m missing something, but I don’t feel like I have a good grasp on the reasoning for constants to follow a different style.

**Comment by anikethsaha** [comment]
I do agree that having an option like this may bring changes to consistency but in a sense, it can keep the consistency, as if `ignorePattern` uses a pattern for representing a variable name format, like for constant, use CAPS etc.
the issue I faced personally as I mentioned was for old es5 code where there are no `const` and `let` keywords and according to the rule, initialization is only allowed for `const` when the option is `never`. and we were using uppercases to represent constants. And constant should be initialized when declared. Surely it wouldn't be an issue for es6+ versions.
